deepchem/models/torch_models/gan.py

```diff
@@ -869,3 +869,116 @@ def predict_gan_generator(self,
+    advantage over conventional GANs, where the loss does not directly convey
+    information about the quality of the model.
+
+    The theory WGANs are based on requires the discriminator's gradient to be
+    bounded.  The original paper achieved this by clipping its weights.  This
+    class instead does it by adding a penalty term to the discriminator's loss, as
+    described in [2].  This is sometimes found to produce better results.
+
+    There are a few other practical differences between GANs and WGANs.  In a
+    conventional GAN, the discriminator's output must be between 0 and 1 so it can
+    be interpreted as a probability.  In a WGAN, it should produce an unbounded
+    output that can be interpreted as a distance.
+
+    When training a WGAN, you also should usually use a smaller value for
+    generator_steps.  Conventional GANs rely on keeping the generator and
+    discriminator "in balance" with each other.  If the discriminator ever gets
+    too good, it becomes impossible for the generator to fool it and training
+    stalls.  WGANs do not have this problem, and in fact the better the
+    discriminator is, the easier it is for the generator to improve.  It therefore
+    usually works best to perform several training steps on the discriminator for
+    each training step on the generator.
+
+    References
+    ----------
+    [1] Arjovsky, Martin, Soumith Chintala, and Léon Bottou.
+        "Wasserstein generative adversarial networks."
+        International conference on machine learning. PMLR, 2017.
+        (https://arxiv.org/abs/1701.07875)
+
+    [2] Gulrajani, Ishaan, et al. "Improved training of wasserstein gans."
+        Advances in neural information processing systems 30 (2017).
+        (https://arxiv.org/abs/1704.00028)
+    """
+
+    def __init__(self, gradient_penalty=10.0, **kwargs):
+        """Construct a WGAN.
+
+        In addition to the following, this class accepts all the keyword arguments
+        from GAN and KerasModel.
+
+        Parameters
+        ----------
+        gradient_penalty: float default 10.0
+            the magnitude of the gradient penalty loss
+        """
+        self.gradient_penalty = gradient_penalty
+        super(WGANModel, self).__init__(**kwargs)
+
+    def _call_discriminator(self, discriminator, inputs, train):
+        if train:
+            penalty = GradientPenaltyLayer(self, discriminator)
+            return penalty(inputs, self.conditional_input_layers)
+        return discriminator(
+            _list_or_tensor(inputs + self.conditional_input_layers))
+
+    def create_generator_loss(self, discrim_output):
+        return torch.mean(discrim_output)
+
+    def create_discriminator_loss(self, discrim_output_train,
+                                  discrim_output_gen):
+        return torch.mean(discrim_output_train[0][0] -
+                          discrim_output_gen[1]) + discrim_output_train[1]
+
+
+class GradientPenaltyLayer(nn.Module):
```

Comment:
More detailed docstrings and usage examples

Reply:
done